File: 402-remove-k-digits/remove-k-digits.php
class Solution {

    /**
     * @param String $num
     * @param Integer $k
     * @return String
     */
    function removeKdigits($num, $k) {
        $s = [];
        foreach (str_split($num) as $d) {
            while ($k > 0 && !empty($s) && end($s) > $d) {
                array_pop($s);
                $k--;
            }
            $s[] = $d;
        }

        while ($k > 0) {
            array_pop($s);
            $k--;
        }

        while (!empty($s) && $s[0] == '0') {
            array_shift($s);
        }

        return empty($s) ? '0' : implode('', $s);
    }
}

File: 402-remove-k-digits/remove-k-digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        st = []
        for d in num:
            while k > 0 and st and st[-1] > d:
                st.pop()
                k -= 1
            st.append(d)
        
        while k > 0:
            st.pop()
            k -= 1
        
        while st and st[0] == '0':
            st.pop(0)
        
        return '0' if not st else ''.join(st)